src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp

```diff
@@ -68,6 +72,31 @@ void BarrierSetAssembler::load_at(MacroAssembler* masm, DecoratorSet decorators,
   }
 }
 
+void BarrierSetAssembler::nmethod_entry_barrier(MacroAssembler* masm, Register tmp) {
+  BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+  if (bs_nm == nullptr) {
+    return;
+  }
+
+  assert_different_registers(tmp, Z_R1_scratch);
+
+  __ block_comment("nmethod_entry_barrier (nmethod_entry_barrier) {");
+
+    // Load jump addr:
+    __ load_const(Z_R1_scratch, (uint64_t)StubRoutines::zarch::nmethod_entry_barrier()); // 2*6 bytes
+
+    // Load value from current java object:
+    __ z_lg(tmp, in_bytes(bs_nm->thread_disarmed_offset()), Z_thread); // 6 bytes
+
+    // Compare to current patched value:
+    __ z_cfi(tmp, /* to be patched */ -1); // 6 bytes (2 + 4 byte imm val)
+
+    // Conditional Jump
+    __ z_bcr(Assembler::bcondNotEqual, Z_R1_scratch); // 2 bytes
```

Comment:
This is only a jump. We would need a call which sets Z_R14 = return address.

Reply:
If this isn't nerdy enough, I have another idea (works only on z14 and newer). It replaces the load_const at the beginning with a z_larl() and then uses a z_bic() to fetch the target address and branch there. The z_larl will never go out of range because all generated code is in the code cache which is limited to 4GB in size. In total we would get: 
```
__ z_larl(Z_R1_scratch, &StubRoutines::zarch::nmethod_entry_barrier());
__ z_clfhsi(in_bytes(bs_nm->thread_disarmed_offset()), Z_thread, /* to be patched */ -1);
__ z_larl(Z_R14, (instr_len((unsigned long)LARL_ZOPC) + instr_len((unsigned long)BCR_ZOPC)) / 2);
__ z_bic(Assembler::bcondNotEqual, 0, Z_R0, Z_R1_scratch);
```
Nice, short, compact.

Reply:
Well spotted. I hadn't appreciated that this instruction wasn't setting the return address.

I am leaning towards the first ~nerdy~ cool hack if it's more general, though I do appreciate the compactness of the second suggestion.

Reply:
Yes, my "optimisations" probably will be of no use because a) the compare value has. to be 4 bytes and b) the big instruction is available from z14 only. 
Anyway, you get a glimpse of the beauty of z programming. :-)

Reply:
The branch target is constant once the stub is created. We don't need to load it from memory.

Reply:
Good idea! One instruction instead of two, plus avoiding a data interlock when loading the constant address.

Reply:
I made this change. Though I do want to confirm that `__ z_larl(Z_R1_scratch, StubRoutines::zarch::nmethod_entry_barrier());` is the correct version, and not `&StubRoutines::zarch::nmethod_entry_barrier()` which is also mentioned above. Since nmethod_entry_barrier returns an `address` (and address is already a `char*`) I don't believe we want the ampersand.

Reply:
You are right. We don't want the '&'. I first assumed (disregarding the parentheses) this would be a field where the address of a generated stub is stored.

Reply:
I think I had the same thought when I wrote the code initially, so I was doubting my reasoning. Thanks for confirming :-)

Reply:
I agree: stick with the load_const as you suggested. 
You can't calculate the distance statically. The generated nmethod is copied around at least once when code generation is complete. You would need a relocation to handle that, and we don't have one. 
Sorry for all the discussion which led to just nothing.

Reply:
No need to apologize! I appreciate the suggestions either way.